service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,538 @@
+		c.logger.Warn("", tag.LifeCycleStopTimedout)
+	}
+
+	c.redispatcher.Stop()
+}
+
+func (c *crossClusterQueueProcessorBase) processorPump() {
+	defer c.shutdownWG.Done()
+
+	updateAckTimer := time.NewTimer(backoff.JitDuration(
+		c.options.UpdateAckInterval(),
+		c.options.UpdateAckIntervalJitterCoefficient(),
+	))
+	defer updateAckTimer.Stop()
+
+	maxPollTimer := time.NewTimer(backoff.JitDuration(
+		c.options.MaxPollInterval(),
+		c.options.MaxPollIntervalJitterCoefficient(),
+	))
+	defer maxPollTimer.Stop()
+
+processorPumpLoop:
+	for {
+		select {
+		case <-c.shutdownCh:
+			break processorPumpLoop
+		case <-c.notifyCh:
+			c.notifyAllQueueCollections()
+		case <-updateAckTimer.C:
+			processFinished, err := c.updateAckLevel()
+			if err == shard.ErrShardClosed || (err == nil && processFinished) {
+				go c.Stop()
+				break processorPumpLoop
+			}
+			updateAckTimer.Reset(backoff.JitDuration(
+				c.options.UpdateAckInterval(),
+				c.options.UpdateAckIntervalJitterCoefficient(),
+			))
+		case <-maxPollTimer.C:
+			c.notifyAllQueueCollections()
+			maxPollTimer.Reset(backoff.JitDuration(
+				c.options.MaxPollInterval(),
+				c.options.MaxPollIntervalJitterCoefficient(),
+			))
+		case <-c.processCh:
+			maxRedispatchQueueSize := c.options.MaxRedispatchQueueSize()
+			if c.redispatcher.Size() > maxRedispatchQueueSize {
+				// has too many pending tasks in re-dispatch queue, block loading tasks from persistence
+				c.redispatcher.Redispatch(maxRedispatchQueueSize)
+				if c.redispatcher.Size() > maxRedispatchQueueSize {
+					// if redispatcher still has a large number of tasks
+					// this only happens when system is under very high load
+					// we should backoff here instead of keeping submitting tasks to task processor
+					time.Sleep(backoff.JitDuration(
+						c.options.PollBackoffInterval(),
+						c.options.PollBackoffIntervalJitterCoefficient(),
+					))
+				}
+				// re-enqueue the event to see if we need keep re-dispatching or load new tasks from persistence
+				select {
+				case c.processCh <- struct{}{}:
+				default:
+				}
+				continue processorPumpLoop
+			}
+			c.processQueueCollections()
+		case notification := <-c.actionNotifyCh:
+			c.handleActionNotification(notification)
+		}
+	}
+}
+
+func (c *crossClusterQueueProcessorBase) readTasks(
+	readLevel task.Key,
+	maxReadLevel task.Key,
+) ([]*persistence.CrossClusterTaskInfo, bool, error) {
+
+	var response *persistence.GetCrossClusterTasksResponse
+	op := func() error {
+		var err error
+		response, err = c.shard.GetExecutionManager().GetCrossClusterTasks(
+			context.Background(),
+			&persistence.GetCrossClusterTasksRequest{
+				TargetCluster: c.targetCluster,
+				ReadLevel:     readLevel.(transferTaskKey).taskID,
+				MaxReadLevel:  maxReadLevel.(transferTaskKey).taskID,
+				BatchSize:     c.options.BatchSize(),
+			})
+		return err
+	}
+
+	err := backoff.Retry(op, persistenceOperationRetryPolicy, persistence.IsTransientError)
```

Comment:
Probably, we should retry on ServiceBusyError. If the error is from database, retry can create more requests which will lead to more errors.

Reply:
Yes. Remove this error from retry policy. The queue processor is going to retry with a lower rate